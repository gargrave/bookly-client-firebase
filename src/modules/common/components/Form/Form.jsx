// @flow
import React from 'react'
import { any, array, func, bool, string } from 'prop-types'

import { buildClass } from '../../../../utils/cssHelpers'

import Alert from '../Alert/Alert'
import Button from '../Button'
import ButtonRow from '../ButtonRow/ButtonRow'

import styles from './Form.css'

type Props = {
  cancelBtnText?: string,
  children?: any,
  classes?: string[],
  disabled?: boolean,
  onCancel?: Function,
  onSubmit: Function,
  submitBtnText?: string,
  submitDisabled?: boolean,
  topLevelError?: string,
}

const Form = ({
  cancelBtnText,
  children,
  classes = [],
  disabled,
  onCancel,
  onSubmit,
  submitBtnText,
  submitDisabled = false,
  topLevelError,
}: Props) => {
  return (
    <div>
      {topLevelError && <Alert message={topLevelError} type="danger" />}
      <form
        className={buildClass(styles.form, ...(classes || []))}
        onSubmit={onSubmit}
        noValidate
      >
        {children}

        <ButtonRow>
          <Button
            canSubmit={true}
            disabled={submitDisabled || disabled || false}
            onClick={onSubmit}
            position="left"
            text={submitBtnText || 'Submit'}
            type="success"
          />

          {onCancel && (
            <Button
              onClick={onCancel}
              position="right"
              text={cancelBtnText || 'Cancel'}
              type="light"
            />
          )}
        </ButtonRow>
      </form>
    </div>
  )
}

Form.propTypes = {
  cancelBtnText: string,
  children: any,
  classes: array,
  disabled: bool,
  onCancel: func,
  onSubmit: func.isRequired,
  submitBtnText: string,
  submitDisabled: bool,
  topLevelError: string,
}

export default Form
