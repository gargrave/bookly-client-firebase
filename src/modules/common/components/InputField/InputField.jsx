// @flow
import React from 'react'
import { bool, func, number, oneOf, string } from 'prop-types'
import styled from 'react-emotion'

import { colors } from '../../../../styles'
import { clamp } from '../../../../utils/mathHelpers'

type Props = {
  boundValue: string,
  disabled?: boolean,
  error?: string,
  label?: string,
  maxLength?: number,
  name: string,
  onInputChange: Function,
  placeholder?: string,
  type?: string,
}

const acceptableTypes = ['email', 'password', 'search', 'text']

const StyledInput = styled('input')`
  border-color: ${props => props.invalid && colors.error} !important;
`

const StyledError = styled('p')`
  color: ${colors.error};
`

const InputField = ({
  boundValue,
  disabled,
  error,
  label,
  maxLength = 255,
  name,
  onInputChange,
  placeholder,
  type,
}: Props) => {
  return (
    <div className="input-field">
      {label && <label htmlFor={name}>{label}:</label>}

      <StyledInput
        disabled={disabled || false}
        id={name}
        invalid={!!error}
        maxLength={clamp(maxLength, 1, 255)}
        name={name}
        onChange={onInputChange}
        placeholder={placeholder || ''}
        type={type || 'text'}
        value={boundValue}
      />

      {error && <StyledError className="error">{error}</StyledError>}
    </div>
  )
}

InputField.propTypes = {
  boundValue: string.isRequired,
  disabled: bool,
  error: string,
  maxLength: number,
  label: string,
  name: string.isRequired,
  onInputChange: func.isRequired,
  placeholder: string,
  type: oneOf(acceptableTypes),
}

export default InputField
