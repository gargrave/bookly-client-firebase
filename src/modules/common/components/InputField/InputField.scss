@import '../../../../styles/_colors';

.error {
  color: $errorColor;
}

input.invalid {
  border-color: $errorColor;
}
