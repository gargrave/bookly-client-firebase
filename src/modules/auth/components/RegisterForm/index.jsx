// @flow
import React from 'react';
import { bool, func, shape, string } from 'prop-types';

import type { RegisterErrors, RegisterUser } from '../../../../modules/auth/flowtypes';

import Form from '../../../../modules/common/components/Form/Form';
import InputField from '../../../../modules/common/components/InputField/InputField';

type Props = {
  disabled?: boolean,
  errors: RegisterErrors,
  registerUser: RegisterUser,
  onInputChange: Function,
  onSubmit: Function,
  submitBtnText?: string,
  submitDisabled?: boolean,
  topLevelError?: string,
};

const RegisterPage = ({
  disabled = false,
  errors,
  registerUser,
  onInputChange,
  onSubmit,
  submitBtnText,
  submitDisabled,
  topLevelError,
}: Props) => (
  <Form
    classes={['register-form']}
    disabled={disabled}
    onSubmit={onSubmit}
    submitBtnText={submitBtnText}
    submitDisabled={submitDisabled}
    topLevelError={topLevelError}
  >
    <InputField
      boundValue={registerUser.email}
      disabled={disabled}
      error={errors.email}
      label="Email"
      name="email"
      onInputChange={onInputChange}
      type="email"
    />

    <InputField
      boundValue={registerUser.password}
      disabled={disabled}
      error={errors.password}
      label="Password"
      name="password"
      onInputChange={onInputChange}
      type="password"
    />

    <InputField
      boundValue={registerUser.passwordConfirm}
      disabled={disabled}
      error={errors.passwordConfirm}
      label="Confirm Password"
      name="passwordConfirm"
      onInputChange={onInputChange}
      type="password"
    />
  </Form>
);

RegisterPage.propTypes = {
  disabled: bool,
  errors: shape({
    email: string.isRequired,
    password: string.isRequired,
    passwordConfirm: string.isRequired,
  }).isRequired,
  registerUser: shape({
    email: string.isRequired,
    password: string.isRequired,
    passwordConfirm: string.isRequired,
  }).isRequired,
  onInputChange: func.isRequired,
  onSubmit: func.isRequired,
  submitBtnText: string,
  submitDisabled: bool,
  topLevelError: string,
};

export default RegisterPage;
