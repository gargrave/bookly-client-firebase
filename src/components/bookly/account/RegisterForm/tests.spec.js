import React from 'react';
import { shallow } from 'enzyme';

import Form from '../../../common/Form';
import InputField from '../../../common/InputField';

import RegisterForm from './';

const defaultProps = Object.freeze({
  disabled: false,
  errors: {
    email: '',
    password: '',
    passwordConfirm: '',
  },
  registerUser: {
    email: 'whatever@gmail.com',
    password: 'password',
    passwordConfirm: 'password',
  },
  onInputChange: jest.fn(),
  onSubmit: jest.fn(),
  submitDisabled: false,
  topLevelError: '',
});

function getComponent(extraProps = {}) {
  const props = Object.assign({}, defaultProps, extraProps);
  return shallow(<RegisterForm {...props} />);
}

describe('RegisterForm', () => {
  let component;

  test('matches the snapshot', () => {
    component = getComponent();
    expect(component).toMatchSnapshot();
  });

  describe('basic rendering', () => {
    test('renders correctly', () => {
      component = getComponent();
      expect(component.find(Form)).toHaveLength(1);
      expect(component.find(InputField)).toHaveLength(3);
    });
  });
});
