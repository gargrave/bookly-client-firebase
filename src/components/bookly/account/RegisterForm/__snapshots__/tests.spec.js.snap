// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegisterForm matches the snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <RegisterPage
    disabled={false}
    errors={
      Object {
        "email": "",
        "password": "",
        "passwordConfirm": "",
      }
    }
    onInputChange={[Function]}
    onSubmit={[Function]}
    registerUser={
      Object {
        "email": "whatever@gmail.com",
        "password": "password",
        "passwordConfirm": "password",
      }
    }
    submitDisabled={false}
    topLevelError=""
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <InputField
          boundValue="whatever@gmail.com"
          disabled={false}
          error=""
          label="Email"
          name="email"
          onInputChange={[Function]}
          type="email"
        />,
        <InputField
          boundValue="password"
          disabled={false}
          error=""
          label="Password"
          name="password"
          onInputChange={[Function]}
          type="password"
        />,
        <InputField
          boundValue="password"
          disabled={false}
          error=""
          label="Confirm Password"
          name="passwordConfirm"
          onInputChange={[Function]}
          type="password"
        />,
      ],
      "classes": Array [
        "register-form",
      ],
      "disabled": false,
      "onSubmit": [Function],
      "submitBtnText": undefined,
      "submitDisabled": false,
      "topLevelError": "",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "boundValue": "whatever@gmail.com",
          "disabled": false,
          "error": "",
          "label": "Email",
          "name": "email",
          "onInputChange": [Function],
          "type": "email",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "boundValue": "password",
          "disabled": false,
          "error": "",
          "label": "Password",
          "name": "password",
          "onInputChange": [Function],
          "type": "password",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "boundValue": "password",
          "disabled": false,
          "error": "",
          "label": "Confirm Password",
          "name": "passwordConfirm",
          "onInputChange": [Function],
          "type": "password",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <InputField
            boundValue="whatever@gmail.com"
            disabled={false}
            error=""
            label="Email"
            name="email"
            onInputChange={[Function]}
            type="email"
          />,
          <InputField
            boundValue="password"
            disabled={false}
            error=""
            label="Password"
            name="password"
            onInputChange={[Function]}
            type="password"
          />,
          <InputField
            boundValue="password"
            disabled={false}
            error=""
            label="Confirm Password"
            name="passwordConfirm"
            onInputChange={[Function]}
            type="password"
          />,
        ],
        "classes": Array [
          "register-form",
        ],
        "disabled": false,
        "onSubmit": [Function],
        "submitBtnText": undefined,
        "submitDisabled": false,
        "topLevelError": "",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "boundValue": "whatever@gmail.com",
            "disabled": false,
            "error": "",
            "label": "Email",
            "name": "email",
            "onInputChange": [Function],
            "type": "email",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "boundValue": "password",
            "disabled": false,
            "error": "",
            "label": "Password",
            "name": "password",
            "onInputChange": [Function],
            "type": "password",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "boundValue": "password",
            "disabled": false,
            "error": "",
            "label": "Confirm Password",
            "name": "passwordConfirm",
            "onInputChange": [Function],
            "type": "password",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
